### PR TITLE
Reporting APIs

### PR DESCRIPTION
**Pull request by k88hudson**
This includes reporting apis and queries for `/stats`.

### CODE REVIEW

**Comment by k88hudson** [comment]
## `GET /stats`

### Query string filters

Filter  | Default Value | Notes
------------- | ------------- | ---------
`group_by`  | `['date']` | An array of fields by which data should be aggregated. Note that when you are encoding this in a query string you should use the form `group_by=date&group_by=category` etc.
`campaign_id` | | (e.g. `12`)
`start_date` | | Search for stats after a date (e.g. `2015-01-01`)
`end_date` | | Search for stats before a date (e.g. `2015-01-01`)
`locale` | all | (e.g. `'en-US'`)
`country_code` | all | (e.g. `'US'`)
`channel_id`| all | This corresponds to the channel, like "desktop" or "prerelease" (e.g. `1`)
`adgroup_type` | all | Must be one of `'suggested'` or `'directory'`
`type` | all | This is stored at the tile level. Must be one of `'affliate'` or `'sponsored'`

**Comment by tkiethanom** [comment]
What are the valid values of group_by?

**Comment by k88hudson** [comment]
@tkiethanom  `'date'`, `'week'`, `'month'`, `'category'`, `'country_code'`, `'locale'`

**Comment by k88hudson** [comment]
All of this is flexible of course ^_^

File: splice/queries/reporting.py [hunk 1]
```diff
@@ -1,33 +1,88 @@
...
-from splice.queries.tile import get_tiles_by_campaign
+from splice.queries.tile import get_tiles
 from sqlalchemy.sql import func
+from sqlalchemy.orm import aliased
+from sqlalchemy import case
 
 
-def get_stats_by_date(campaign_id):
-
+def get_stats(group_by, filters=None):
```
Comment (ncloudioj) [inline splice/queries/reporting.py]: Add some doc strings here to outline the function and arguments could be helpful for the API users.

File: splice/queries/tile.py [hunk 2]
```diff
@@ -1,45 +1,67 @@
...
-
-    rows = (
-        env.db.session
-        .query(Tile)
-        .filter(Tile.adgroup_id == adgroup_id)
-        .order_by(Tile.id.desc())
-        .all()
-    )
-    output = [row_to_dict(d) for d in rows]
+def needsAdgroupJoin(filters):
```
Comment (ncloudioj) [inline splice/queries/tile.py]: favor `needs_adgroup_join` in Python :)

File: splice/queries/tile.py [hunk 3]
```diff
@@ -1,45 +1,67 @@
...
+    adgroup_filters = ('adgroup_type', 'channel_id', 'locale')
+    return all(k in filters for k in adgroup_filters)
```
Comment (ncloudioj) [inline splice/queries/tile.py]: Perhaps this is more straightforward,
```python
_AdgroupFilters = set(['adgroup_type', 'channel_id', 'locale'])

def needs_adgroup_join(filters):
    return _AdgroupFilters.issubset(filters)
```

File: splice/queries/tile.py [hunk 4]
```diff
@@ -1,45 +1,67 @@
...
+        rows = (
+            rows
+            .select_from(Tile)
+            .join(AdgroupCategory, Tile.adgroup_id == AdgroupCategory.adgroup_id)
+            .join(Adgroup, AdgroupCategory.adgroup_id == Adgroup.id)
+            .group_by(AdgroupCategory.category)
+        )
+    else:
+        rows = rows.order_by(Tile.id.desc())
+    if (campaign_id or needsAdgroupJoin(filters)) and not group_by == 'category':
```
Comment (ncloudioj) [inline splice/queries/tile.py]:
```python
if campaign_id or needsAdgroupJoin(filters) and group_by != 'category':
```

File: splice/web/api/reporting.py [hunk 5]
```diff
@@ -1,34 +1,75 @@
...
-    'blocked': fields.Integer
+    'week': fields.Integer,
+    'month': fields.Integer,
+    'category': fields.String,
+    'locale': fields.String,
+    'country_code': fields.String
 }
 
+valid_filters = {
+    'type': {'choices': ('affiliate', 'sponsored'), 'store_missing': False},
```
Comment (ncloudioj) [inline splice/web/api/reporting.py]: A constant "type" has been defined in the `models.Tiles` class, we can use that here. The same to the "adgroup_type" below.

File: splice/web/api/reporting.py [hunk 6]
```diff
@@ -1,34 +1,75 @@
...
+    'adgroup_type': {'choices': ('suggested', 'directory'), 'store_missing': False},
+    'country_code': {'store_missing': False},
+    'locale': {'store_missing': False},
+    'channel_id': {'type': int, 'store_missing': False},
+    'campaign_id': {'type': int, 'store_missing': False},
+    'start_date': {'store_missing': False},
```
Comment (ncloudioj) [inline splice/web/api/reporting.py]: There is a dedicated type for the `start_dat`e and `end_date`, it'll handle the data validation for you.

```python
from flask-restful import inputs

type=inputs.date
```

File: splice/queries/tile.py [hunk 7]
```diff
@@ -1,45 +1,67 @@
...
-    )
-    output = [row_to_dict(d) for d in rows]
+def needsAdgroupJoin(filters):
+    adgroup_filters = ('adgroup_type', 'channel_id', 'locale')
+    return all(k in filters for k in adgroup_filters)
 
-    return output
 
+def get_tiles(campaign_id=None, adgroup_id=None, group_by=None,
+              limit_fields=None, filters=None):
```
Comment (ncloudioj) [inline splice/queries/tile.py]: :+1: super powerful Swiss-Knife function. Takes time to walk through it though.

Perhaps we can refactor it a bit for the readability, like break it down to a few sub-functions, base query builder, then add join tables and group by columns, finally inject filters.

File: splice/queries/tile.py (hunk 2 above)
Comment (k88hudson) [inline splice/queries/tile.py]: oops! :stuck_out_tongue_winking_eye:

File: splice/queries/tile.py (hunk 3 above)
Comment (k88hudson) [inline splice/queries/tile.py]: ah, cool! thanks

**Comment by ncloudioj** [comment]
Awesome sauce :clap:. Looks good to me except some improvements for the readability.

File: splice/queries/tile.py (hunk 7 above)
Comment (k88hudson) [inline splice/queries/tile.py]: agreed, the reporting query could use some breaking down as well

**Comment by ncloudioj** [comment]
r+!

**Comment by k88hudson** [comment]
@ncloudioj i refactored that function a bit, want to take a look?

**Comment by ncloudioj** [comment]
@k88hudson Much better now. Merge it!

**Pull request merged**